src/main/utility/sockaddr.rs

```diff
@@ -0,0 +1,688 @@
+            return None;
+        }
+
+        SockaddrUnix::new(unsafe { &mut self.addr.unix }, self.len)
+    }
+
+    /// Get a new `SockaddrStorage` with a copy of the unix socket address.
+    pub fn from_unix(addr: &SockaddrUnix<&libc::sockaddr_un>) -> Self {
+        let (ptr, len) = addr.as_ptr();
+
+        unsafe { Self::from_ptr(ptr as *const MaybeUninit<u8>, len) }.unwrap()
+    }
+
+    /// A pointer to the socket address. Some bytes may be uninitialized.
+    pub fn as_ptr(&self) -> (*const MaybeUninit<u8>, libc::socklen_t) {
+        (unsafe { &self.addr.slice }.as_ptr(), self.len)
+    }
+
+    /// The socket address as a slice of bytes. Some bytes may be uninitialized.
+    pub fn as_slice(&self) -> &[MaybeUninit<u8>] {
+        unsafe { &self.addr.slice[..(self.len as usize)] }
+    }
+
+    /// Returns true if the address length is large enough for a `AsType`, and if it has the
+    /// provided family.
+    fn verify_fixed_size<AsType>(&self, family: AddressFamily) -> bool {
+        (self.len as usize) >= std::mem::size_of::<AsType>() && self.family() == Some(family)
+    }
+}
+
+impl std::fmt::Debug for SockaddrStorage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SockaddrStorage")
+            .field("family", &self.family())
+            .field("len", &self.len)
+            .finish_non_exhaustive()
+    }
+}
+
+/// A Unix socket address. Typically will be used as an owned address
+/// `SockaddrUnix<libc::sockaddr_un>` or a borrowed address `SockaddrUnix<&libc::sockaddr_un>`, and
+/// you can convert between them using methods such as [`as_ref`](Self::as_ref) or
+/// [`into_owned`](Self::into_owned).
+#[derive(Clone, Copy)]
+pub struct SockaddrUnix<T>
+where
+    T: Borrow<libc::sockaddr_un>,
+{
+    addr: T,
+    len: libc::socklen_t,
+}
+
+impl<T> SockaddrUnix<T>
+where
+    T: Borrow<libc::sockaddr_un>,
+{
+    /// Get a new `SockaddrUnix` for a `libc::sockaddr_un`. The `libc::sockaddr_un` must be
+    /// properly initialized. Will return `None` if the length is too short, too long, or if
+    /// `sun_family` is not `AF_UNIX`.
+    pub fn new(addr: T, len: libc::socklen_t) -> Option<Self> {
+        if (len as usize) < memoffset::span_of!(libc::sockaddr_un, sun_family).end {
+            return None;
+        }
+
+        if (len as usize) > std::mem::size_of::<libc::sockaddr_un>() {
+            return None;
+        }
+
+        if addr.borrow().sun_family as i32 != libc::AF_UNIX {
+            return None;
+        }
+
+        Some(Self { addr, len })
+    }
+
+    /// If the socket address represents a pathname address, returns the C string representing the
+    /// filesystem path.
+    pub fn as_path(&self) -> Option<&CStr> {
+        let path = self.sun_path()?;
+
+        if path.is_empty() || path[0] == 0 {
+            return None;
+        }
+
+        let first_nul = path.iter().position(|&x| x == 0)?;
```

Comment:
Should we care if the NULL occurs earlier than expected based on the length?

Reply:
For abstract named sockets, the length is important since the address name isn't a string and doesn't have a nul terminator, so the length directly affects the socket address. For pathname sockets, the length doesn't matter as long as it's long enough to include the string and nul terminator. So it's common for application code to just write their string to `sun_path` and in the syscall just use `sizeof(sockaddr_un)` as the length (and Linux allows this).

Reply:
I see - a comment here along those lines would be helpful

Reply:
Added a comment.